sql/core/src/test/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListenerSuite.scala

```diff
@@ -142,288 +155,274 @@ class SQLListenerSuite extends SparkFunSuite with SharedSQLContext with JsonTest
       jobId = 0,
       time = System.currentTimeMillis(),
       stageInfos = Seq.empty,
       createProperties(executionId)))
-    bus.postToAll(SparkListenerSQLExecutionEnd(
+    listener.onOtherEvent(SparkListenerSQLExecutionEnd(
       executionId, System.currentTimeMillis()))
-    bus.postToAll(SparkListenerJobEnd(
+    listener.onJobEnd(SparkListenerJobEnd(
       jobId = 0,
       time = System.currentTimeMillis(),
       JobFailed(new RuntimeException("Oops"))
     ))
 
-    assertJobs(store.execution(0), failed = Seq(0))
+    assertJobs(statusStore.execution(executionId), failed = Seq(0))
   }
 
   test("SPARK-11126: no memory leak when running non SQL jobs") {
-    val previousStageNumber = statusStore.executionsList().size
+    val statusStore = createStatusStore()
+    val listener = statusStore.listener.get
+    // No need to remove the listener as the SparkContext is dedicated to this test suite.
+    sparkContext.addSparkListener(listener)
     spark.sparkContext.parallelize(1 to 10).foreach(i => ())
     spark.sparkContext.listenerBus.waitUntilEmpty(10000)
-    // listener should ignore the non SQL stage
-    assert(statusStore.executionsList().size == previousStageNumber)
-
-    spark.sparkContext.parallelize(1 to 10).toDF().foreach(i => ())
-    spark.sparkContext.listenerBus.waitUntilEmpty(10000)
-    // listener should save the SQL stage
-    assert(statusStore.executionsList().size == previousStageNumber + 1)
+    // Listener should ignore the non-SQL stages, as the stage data are only removed when SQL
+    // execution ends, which will not be triggered for non-SQL jobs.
+    assert(listener.stageMetrics.isEmpty)
   }
 
   test("driver side SQL metrics") {
-    val oldCount = statusStore.executionsList().size
-    val expectedAccumValue = 12345L
+    val statusStore = createStatusStore()
+    val listener = statusStore.listener.get
+    // No need to remove the listener as the SparkContext is dedicated to this test suite.
+    sparkContext.addSparkListener(listener)
+    val expectedAccumValue = 12345
     val physicalPlan = MyPlan(sqlContext.sparkContext, expectedAccumValue)
     val dummyQueryExecution = new QueryExecution(spark, LocalRelation()) {
       override lazy val sparkPlan = physicalPlan
       override lazy val executedPlan = physicalPlan
     }
-
     SQLExecution.withNewExecutionId(spark, dummyQueryExecution) {
       physicalPlan.execute().collect()
     }
 
-    while (statusStore.executionsList().size < oldCount) {
-      Thread.sleep(100)
-    }
-
-    // Wait for listener to finish computing the metrics for the execution.
-    while (statusStore.executionsList().last.metricValues == null) {
+    // Wait until execution finished
+    var finished = false
+    while (!finished) {
       Thread.sleep(100)
+      val executionData = statusStore.executionsList().headOption
+      finished = executionData.isDefined && executionData.get.jobs.values
+        .forall(_ == JobExecutionStatus.SUCCEEDED)
```

Comment:
This is not the same check as before. It assumes that `onJobEnd` is called after `onExecutionEnd`, which might be the case now, but was explicitly mentioned as not being always the case in the original `SQLListener` code.